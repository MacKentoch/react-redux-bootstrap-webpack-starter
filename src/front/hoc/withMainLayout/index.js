// @flow

import withMainLayout from './withMainLayout';

export default withMainLayout;
