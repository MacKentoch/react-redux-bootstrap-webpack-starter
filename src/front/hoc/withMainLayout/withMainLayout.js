// @flow

// #region imports
import React, { Component } from 'react';
import wrapDisplayName from 'recompose/wrapDisplayName';
import compose from 'recompose/compose';
import { withRouter } from 'react-router';
import { type Match, type Location, type RouterHistory } from 'react-router';
import NavigationBar from '../../components/navigation';
import BackToTop from '../../components/backToTop/BackToTop';
import navigationModel from '../../config/navigation';
import { type Navigation } from '../../config/navigation';
import registerServiceWorker from '../../services/sw/registerServiceWorker';
// #endregion

// #region flow types
type Props = {
  // from withRouter HOC:
  match: Match,
  location: Location,
  history: RouterHistory,

  ...any,
};

type State = {
  navModel: Navigation,

  ...any,
};
// #endregion

// #region withMainLayout HOC
function withMainLayout(/* no args option yet, but could pass them here */) {
  return BaseComponent => {
    // #region returned Component
    class WithMainLayout extends Component<Props, State> {
      state = { navModel: navigationModel };

      // #region lifecycle
      componentDidMount() {
        if (typeof window !== undefined) {
          // register service worker (no worry about multiple attempts to register, browser will ignore when already registered)
          registerServiceWorker();
        }
      }

      render() {
        /* eslint-disable no-unused-vars */
        const { history, location, match, ...passProps } = this.props;
        /* eslint-enable no-unused-vars */
        const { navModel } = this.state;

        return (
          <div id="appContainer">
            <NavigationBar
              brand={navModel.brand}
              navModel={navModel}
              handleLeftNavItemClick={this.handleLeftNavItemClick}
              handleRightNavItemClick={this.handleRightNavItemClick}
            />
            <div className="container-fluid">
              <BaseComponent {...passProps} />
            </div>
            <BackToTop minScrollY={40} scrollTo={'appContainer'} />
          </div>
        );
      }
      // #endregion

      /* eslint-disable no-unused-vars*/
      handleLeftNavItemClick = (event: SyntheticEvent<>, viewName: string) => {
        // something to do here?
      };

      handleRightNavItemClick = (event: SyntheticEvent<>, viewName: string) => {
        // something to do here?
      };
      /* eslint-enable no-unused-vars*/
    }

    // #region add static displayName for dev
    /* eslint-disable no-process-env */
    if (process.env.NODE_ENV !== 'production') {
      // HOC would obfuscate component name, this trick is helpful for dev (we don't care in production)
      WithMainLayout.displayName = wrapDisplayName(
        BaseComponent,
        'withMainLayout',
      );
    }
    /* eslint-enable no-process-env */
    // #endregion

    return compose(withRouter)(WithMainLayout);
  };
}
// #endregion

export default withMainLayout;
