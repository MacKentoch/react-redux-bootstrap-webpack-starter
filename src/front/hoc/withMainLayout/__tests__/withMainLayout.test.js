import React from 'react';
import { shallow } from 'enzyme';
import withMainLayout from '../withMainLayout';

describe('withMainLayout HOC', () => {
  it('renders as expected', () => {
    const DummyComponent = () => <p>component</p>;
    const WrappedComponent = withMainLayout()(DummyComponent);

    const component = shallow(<WrappedComponent />);
    expect(component).toMatchSnapshot();
  });
});
