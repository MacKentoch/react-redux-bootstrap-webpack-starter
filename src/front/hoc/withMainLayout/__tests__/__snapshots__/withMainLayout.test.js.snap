// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withMainLayout HOC renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withRouter(withMainLayout(DummyComponent)) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
